pkg/api/types.go

```diff
@@ -2353,6 +2363,20 @@ type ServiceSpec struct {
 	// cloud-provider does not support the feature."
 	// +optional
 	LoadBalancerSourceRanges []string
+
+	// externalTraffic denotes if this Service desires to route external traffic to
+	// local endpoints only. This preserves Source IP and avoids a second hop for
+	// LoadBalancer and Nodeport type services.
+	// +optional
+	ExternalTraffic ServiceExternalTrafficType
```

Comment:
Yeah I have the same thought.
[function defaultExternalTrafficIfNeeded()](https://github.com/kubernetes/kubernetes/pull/41162/commits/744766dd401f8f1fcbafd57142a91938ee5bcfef#diff-8b1ee383fe1e9f26f76c775a274ec900R553) exists for defaulting this field to Global for LoadBalancer and NodePort service if this field is not set.
And [function externalTrafficUpdate()](https://github.com/kubernetes/kubernetes/pull/41162/commits/744766dd401f8f1fcbafd57142a91938ee5bcfef#diff-8b1ee383fe1e9f26f76c775a274ec900R271) exists for changing this field back to "" when the type is updated to others (not LoadBalancer nor NodePort).